dali/pipeline/operators/operator.h

```diff
@@ -159,56 +160,122 @@ class DLL_PUBLIC OperatorBase {
-          Flatten(ws);
-        }
-    }
+  void Run(SupportWorkspace *ws) override {
     CheckInputLayouts(ws, spec_);
     SetupSharedSampleParams(ws);
     for (int i = 0; i < input_sets_; ++i) {
-      if (std::is_same<Backend, GPUBackend>::value) {
-        // Before we start working on the next input set, we need
-        // to wait until the last one is finished. Otherwise for some ops
-        // we risk overwriting data used by the kernel called for previous
-        // image. Doing it for all ops is a compromise between performance
-        // (which should not be greatly affected) and robustness (guarding
-        // against this potential problem for newly added ops)
-        SyncHelper(i, ws);
-      }
       RunImpl(ws, i);
     }
-    if (std::is_same<Backend, GPUBackend>::value) {
-      if (sequences_allowed_ && needs_flattening_) {
-        Unflatten(ws);
-      }
+  }
+
+  /**
+   * @brief Implementation of the operator - to be
+   * implemented by derived ops.
+   */
+  virtual void RunImpl(SupportWorkspace *ws, int idx = 0) = 0;
+
+  /**
+   * @brief Shared param setup
+   */
+  virtual void SetupSharedSampleParams(SupportWorkspace *ws) {}
+};
+
+template <>
+class Operator<CPUBackend> : public OperatorBase {
+ public:
+  inline explicit Operator(const OpSpec &spec) : OperatorBase(spec) {}
+
+  inline ~Operator() noexcept(false) override {}
+
+  using OperatorBase::Run;
+  void Run(SampleWorkspace *ws) override {
+    CheckInputLayouts(ws, spec_);
+    SetupSharedSampleParams(ws);
+    for (int i = 0; i < input_sets_; ++i) {
+      RunImpl(ws, i);
     }
   }
 
+  /**
+   * @brief Legacy implementation of CPU operator using per-sample approach
+   *
+   * Usage of this API will be deprecated.
+   */
+  virtual void RunImpl(SampleWorkspace *ws, int idx = 0) {}
+
+  /**
+   * @brief Implementation of the operator - to be implemented by derived ops.
+   */
+  virtual void RunImpl(HostWorkspace *ws, int idx = 0) {
+    DALI_ENFORCE(false, "Not implemented yet");
+  }
+
+  /**
+   * @brief Shared param setup. Legacy implementation for per-sample approach
+   *
+   * Usage of this API will be deprecated.
+   */
+  virtual void SetupSharedSampleParams(SampleWorkspace *ws) {}
+
   /**
    * @brief Shared param setup
    */
-  virtual void SetupSharedSampleParams(Workspace<Backend> *ws) {}
+  virtual void SetupSharedSampleParams(HostWorkspace *ws) {}
+};
+
+template <>
+class Operator<GPUBackend> : public OperatorBase {
+ public:
+  inline explicit Operator(const OpSpec &spec)
+      : OperatorBase(spec),
+        sequences_allowed_(SchemaRegistry::GetSchema(spec.name()).AllowsSequences()),
+        needs_flattening_(SchemaRegistry::GetSchema(spec.name()).NeedsFlattening()) {}
+
+  inline ~Operator() noexcept(false) override {}
+
+  using OperatorBase::Run;
+  void Run(DeviceWorkspace *ws) override {
+    std::vector<std::vector<int>> seq_sizes;
+    if (sequences_allowed_ && needs_flattening_) {
```

Comment:
this collides with #930, where I got rid of sequence flattening/unflattening

Reply:
Cannot do it unless one is merged. Shouldn't be a problem.

Reply:
I'll merge mine 930 pretty soon I think